apps/alb_monitor/lib/alb_monitor/monitor.ex

```diff
@@ -0,0 +1,100 @@
+defmodule ALBMonitor.Monitor do
+  @moduledoc "See `ALBMonitor`."
+
+  use GenServer
+
+  alias ExAws.ElasticLoadBalancingV2, as: ALB
+
+  require Logger
+
+  @ex_aws Application.compile_env!(:alb_monitor, :ex_aws)
+  @http Application.compile_env!(:alb_monitor, :http)
+
+  defmodule State do
+    @moduledoc "State of the GenServer."
+    defstruct check_interval: 5_000,
+              ecs_metadata_uri: nil,
+              instance_ip: nil,
+              shutdown_fn: &:init.stop/0,
+              target_group_arn: nil
+
+    def default do
+      %__MODULE__{
+        ecs_metadata_uri: Application.get_env(:alb_monitor, :ecs_metadata_uri),
+        target_group_arn: Application.get_env(:alb_monitor, :target_group_arn)
+      }
+    end
+  end
+
+  @spec start_link(%State{}) :: GenServer.on_start()
+  def start_link, do: State.default() |> start_link()
```

Comment:
I'm surprised Credo doesn't complain about the single pipe.
```elixir
def start_link(initial_state // State.default()) do
  GenServer.start_link(__MODULE__, initial_state)
end
```

Reply:
For some reason I was pretty strongly convinced this would call `State.default` at compile time and thus defeat the point, but a quick `IO.puts` experiment proves it doesn't. And, actually, now that I check the Elixir guides, there is this sentence:

> Any expression is allowed to serve as a default value, but it won’t be evaluated during the function definition.

I guess it basically compiles to exactly what I wrote out here. Neat!